### ISSUE TITLE
No custom_sqlite3 bundle?

### ISSUE THREAD

**Issue by glen-nicol** [ericsink/SQLitePCL.raw#104]
Looks like there are only bundles for green, e_sqlite3, and sqlcipher in gen_build.cs. Can one be added for the custom_sqlite3 provider?

**Comment by ericsink** [comment]
Maaaaaybe.

But what would this bundle DO?

I mean, it could bring in the custom_sqlite3 provider for every platform.  And it could provide the Batteries.Init() function to do initialization.

But the other bundles bring in the native SQLite library.  In fact, this is most of the value of these bundles.

bundle_custom_sqlite3 couldn't do that.  So it would only work when somebody *also* provides the proper custom_sqlite3 build.  On each platform.  And properly configured to end up in the output directory in such a way that the custom_sqlite3 provider can find it.  On each platform.  

That native lib stuff is most of the effort of a bundle, so bundle_custom_sqlite3 hasn't saved anybody much time.

(I should write up some docs on how to use the custom_sqlite3 provider.)

A package named bundle_custom_sqlite3 would overpromise and underdeliver, disappointing almost everyone.  So that's why I haven't created it.

Does this make sense?

Feel free to tell me more about what your situation and you had in mind when you asked the question.  Making this bundle might still be a good idea.

**Comment by glen-nicol** [comment]
I am considering forking https://github.com/praeclarum/sqlite-net which references the green bundle. If my fork can reference this proposed "custom" bundle I figured I just had to provide the custom library in the final output directory or include it somewhere along the line of packages.

If my fork can just as easily reference core and then call 
```csharp
SQLitePCL.raw.SetProvider(new SQLitePCL.SQLite3Provider_custom_sqlite3());
```
Rather than 
```charp
SQLitePCL.Batteries.Init();
```
Great! But since the readme was a bit scarce on custom usage I felt that the bundle approach would be safer.

**Comment by ericsink** [comment]
Yeah, the raw.SetProvider() call is the preferred approach for your situation.

And, just because I'm curious:

What platform(s) are you going to be targeting?

How will your custom SQLite3 build be different?

**Comment by glen-nicol** [comment]
Alrighty will do.

Windows and iOS. Version of sqlite has some proprietary extensions under the hood. 

Thanks for the quick help!

**Comment by glen-nicol** [comment]
Actually, I have another question. Will this [provider](https://www.nuget.org/packages/SQLitePCLRaw.provider.custom_sqlite3.net35/) work on iOS and other PCL targets?

I see it lists .net 3.5 but I thought all it needed was the pinvoke names set to custom_sqlite3 which should work on a PCL, right? I am not quite familiar enough with PCL to know if it will work with a 3.5 dependency.

**Comment by ericsink** [comment]
For your purposes, no, a portable version won't work.

iOS is a very special case, because Xamarin is different, and because iOS does not support dynamic linking.  Or rather, it didn't until iOS 8.  And I'm not sure the Xamarin toolchain supports it.

Anyway, for iOS you need to use SQLitePCLRaw.provider.internal.ios_unified, and your custom SQLite3 build needs to be a static library in an embedded resource compiled by Xamarin into an assembly with a LinkWith attribute and also some kind of pointless function you can call just to prevent the Xamarin linker from discarding the assembly.  (Sometimes I love a good run-on sentence.)

The best way to see how this works is to:

- csc gen_build.cs
- ./gen_build.exe
- cd bld
- cd SQLitePCLRaw.lib.e_sqlite3.ios_unified.static

and look at the csproj to see how all this is done.

**Comment by glen-nicol** [comment]
Xamarin does support [dylibs](https://developer.xamarin.com/guides/ios/advanced_topics/native_interop/#Accessing_C_Dylibs). Shouldn't it work as long as I can keep the architecture in sync with the native lib that is deployed?

**Comment by ericsink** [comment]
(Oh cool.  Dylibs.  I'll need to explore that...)

Er, no.  For iOS, even with dynamic loading, [I assume] the pinvokes still need special attributes to marshal the callbacks.

Actually, I suppose for a limited set of SQLite API calls with no callbacks used, what you're describing would work.  But the breadth of stuff supported by SQLitePCL.raw means that iOS providers generally need to be compiled specially for iOS.

So, if you wanted to use dylibs with iOS, I would need to build provider.custom_sqlite3.ios_unified.  This line of gen_build.cs is currently commented out because it would only make sense with dylibs.  But obviously I haven't done any testing with this approach.

Bottom line:  I'm still pretty sure the pinvokes cannot be portable, at least not when iOS is in play.

**Comment by glen-nicol** [comment]
Okay thanks for your help. From my understanding any callbacks need to be marked with [MonoPInvokeCallbackAttribute](https://developer.xamarin.com/api/type/MonoTouch.MonoPInvokeCallbackAttribute/) and have to be static due to the ahead of time compilation for iOS.

I don't currently have to get iOS working because windows is the higher priority so when that time comes I may have to revisit this issue. You may get a pull request from me if you haven't already addressed it.

**Comment by ericsink** [comment]
"Haven't already addressed it"

What does "it" refer to?

If you mean those iOS pinvoke callback attributes, I do already have all that stuff in place. I was just explaining why the pinvokes cannot be portable.
